### DISCUSSION TITLE
What is the OpenTelemetry equivalent for opentracing MockTracer

### DISCUSSION THREAD

**Discussion by srinusanduri** [open-telemetry/opentelemetry-java-instrumentation#7166] [Q&A]
Hello,

OpenTracing has a `io.opentracing.mock.MockTracer` implementation which allows us to test the spans our code generates. After the code is executed, we can retrieve the finished spans from mock tracer and perform assertions on the spans generated.

```
List<MockSpan> mockSpans = mockTracer.finishedSpans();
```

Is there an equivalent mock library in OpenTelemetry which can be used for testing purposes. If not, how can we achieve this testing functionality in OpenTelemetry Java.

**Comment by trask** [comment]
hi @srinusanduri, check out https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk/testing

**Reply by srinusanduri** [reply]
This is exactly what I am looking for. Thank you @trask.

**Reply by srinusanduri** [reply]
@trask https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk/testing works perfect for testing manually instrumented spans. 

Is there any library to verify the spans instrumented by the agent as well? I would like to write integration test cases in consuming service by attaching the agent and then verify the spans the agent generated. For example, in my spring boot integration test case, I would like to attach agent (with `mock` jvm argument to indicate agent to initialize with span exporter something like `InMemorySpanExporter`) and capture the spans the agent generated at the end of the test case. I was searching for functionality similar to `InMemorySpanExporter` which holds all spans generated in sdk testing and lets us retrieve spans at the end to verify. 

I tried to configure agent with `InMemorySpanExporter` in testing mode but couldn't get it working. We need a mechanism to access spans generated by agent within an integration test.

**Reply by trask** [reply]
@srinusanduri check out https://github.com/open-telemetry/opentelemetry-java-docs/pull/47
